qiskit/pulse/transforms/canonicalization.py

```diff
@@ -85,24 +85,31 @@ def compress_pulses(schedules: List[Schedule]) -> List[Schedule]:
     Returns:
         Compressed schedules.
     """
-    existing_pulses = []
+    pulse_index = 0
+    existing_pulse_set = set()
+    pulse_index_map = dict()    # mapping pulse -> index
+    index_pulse_map = dict()    # mapping index -> pulse
     new_schedules = []
 
     for schedule in schedules:
         new_schedule = Schedule.initialize_from(schedule)
 
         for time, inst in schedule.instructions:
             if isinstance(inst, instructions.Play):
-                if inst.pulse in existing_pulses:
-                    idx = existing_pulses.index(inst.pulse)
-                    identical_pulse = existing_pulses[idx]
+                if inst.pulse in existing_pulse_set:
+                    idx = pulse_index_map[inst.pulse]
+                    identical_pulse = index_pulse_map[idx]
+
                     new_schedule.insert(
                         time,
                         instructions.Play(identical_pulse, inst.channel, inst.name),
                         inplace=True,
                     )
                 else:
-                    existing_pulses.append(inst.pulse)
+                    existing_pulse_set.add(inst.pulse)
```

Comment:
The test failure here is because `Waveform.__hash__` is using deprecated functionality (`numpy.ndarray.tostring`).  That's a bug that existed before your PR, but we must not be hitting that code path.  We'll need to fix that before this could merge - you could fix it in this PR if you like, or we can do it in a separate one.